### PR TITLE
updated adafactor doc  #154862

### LINKED ISSUE #154862: documentation of Adafactor does not match the implementation
### 📚 The doc issue

Hi, I have one question about the implementation of Adafactor. 

According to the PyTorch document and the [Adafactor paper](https://arxiv.org/pdf/1804.04235), a `sum` of the gradient square should be used. 
However, the existing implementations use a `mean` of the gradient square. 

The `mean` of the gradient square is used in the implementations.
https://github.com/pytorch/pytorch/blob/v2.7.0/torch/optim/_adafactor.py#L538
https://github.com/pytorch/pytorch/blob/v2.7.0/torch/optim/_adafactor.py#L547
https://github.com/facebookresearch/fairseq/blob/main/fairseq/optim/adafactor.py#L236
https://github.com/google-research/t5x/blob/main/t5x/adafactor.py#L524

The `sum` of the gradient square is used in the PyTorch document and Adafactor paper.
https://docs.pytorch.org/docs/stable/generated/torch.optim.Adafactor.html
https://arxiv.org/pdf/1804.04235

Why is this discrepancy not documented or discussed?

Reference:
https://github.com/pytorch/pytorch/issues/109581

@janeyx99 

### Suggest a potential alternative/fix

_No response_

cc @svekars @sekyondaMeta @AlannaBurke @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by morrison-turnansky**
updated adafactor doc to reflect difference in implementation vs original paper

Fixes #154862

### CODE REVIEW

File: torch/optim/_adafactor.py [hunk 1]
```diff
@@ -315,6 +315,11 @@ def step(self, closure=None):
                 &\hspace{5mm}U_t \leftarrow \frac{G_t}{\sqrt{\widehat{V}_t}}                                            \\
             \end{aligned}
 
+        In addition, there is an additional deviation in our implementaion when compared to the standard procedure.
+        Noam Shazeer and Mitchell Stern use the sum of the square of the gradient. In our implementation,
+        we use the mean of the square of the gradient and are careful to account for
+        the normalization factor. This allows for greater numerical stability for largs sums where we have limited
+        numerical ranges.
```
Comment (janeyx99) [inline torch/optim/_adafactor.py]: To be clear, there is no mathematical deviation. We are computing the same thing, but using mean instead of sum to take advantage of being able to use smaller numbers.

Comment (morrison-turnansky) [reply torch/optim/_adafactor.py]: Thanks for the feedback! I’ve updated the wording to reflect that this is mathematically equivalent and that the use of mean is purely for numerical stability. Let me know if the new phrasing works for you.

Comment (morrison-turnansky) [reply torch/optim/_adafactor.py]: Please let me know if any additional changes are necessary. @albanD @janeyx99

File: torch/optim/_adafactor.py [hunk 2]
```diff
@@ -315,6 +315,11 @@ def step(self, closure=None):
                 &\hspace{5mm}U_t \leftarrow \frac{G_t}{\sqrt{\widehat{V}_t}}                                            \\
             \end{aligned}
 
+        In addition, there is an additional deviation in our implementaion when compared to the standard procedure.
+        Noam Shazeer and Mitchell Stern describe using the sum of squared gradients,
+        this implementation uses the mean instead. This choice is mathematically equivalent because
+        normalization factors are adjusted accordingly. This allows for greater numerical stability for largs sums
+        where we have limited numerical ranges.
```
Comment (janeyx99) [inline torch/optim/_adafactor.py]:
```suggestion
        You may note that Noam Shazeer and Mitchell Stern describe using the sum of squared gradients,
        while this implementation uses the mean instead. This choice is mathematically equivalent and
        allows for greater numerical stability for large sums.
```

**Comment by janeyx99** [comment]
Sorry I have been a little busy with getting things in 2.8 😛 Thanks for your phrasing--I've made more things concise but LGTM now!

**Review by janeyx99 — APPROVED** [review]

**Comment by morrison-turnansky** [comment]
@janeyx99 All good, and thank you!
